File: src/wildos_graphnav_planner/src/planner_node.cpp
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <nav_msgs/msg/path.hpp>
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/header.hpp>
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
#include <tf2_ros/buffer.h>
#include <tf2_ros/transform_listener.h>

#include <optional>

#include "wildos_graphnav_planner/planner.hpp"

namespace wildos_graphnav_planner
{

class PlannerNode : public rclcpp::Node
{
public:
  explicit PlannerNode(const rclcpp::NodeOptions& options)
  : Node("planner_node", options), tf_buffer_(this->get_clock()), tf_listener_(tf_buffer_, this), planner_(this->get_logger())
  {
    declare_parameter("frontier_dist_cost_factor", 2.0);
    declare_parameter("goal_dist_cost_factor", 1.0);
    declare_parameter("frontier_score_factor", 10.0);
    declare_parameter("min_local_frontier_score", 0.4);
    declare_parameter("local_frontier_radius", 7.0);
    declare_parameter("path_smoothness_period", 10.0);
    declare_parameter("trav_class", "default");
    declare_parameter("goal_radius", 3.0);

    planner_.frontier_dist_cost_factor_ = get_parameter("frontier_dist_cost_factor").as_double();
    planner_.goal_dist_cost_factor_ = get_parameter("goal_dist_cost_factor").as_double();
    planner_.frontier_score_factor_ = get_parameter("frontier_score_factor").as_double();
    planner_.min_local_frontier_score_ = get_parameter("min_local_frontier_score").as_double();
    planner_.local_frontier_radius_ = get_parameter("local_frontier_radius").as_double();
    planner_.path_smoothness_period_ = get_parameter("path_smoothness_period").as_double();
    planner_.set_trav_class(get_parameter("trav_class").as_string());
    goal_radius_ = get_parameter("goal_radius").as_double();

    graph_sub_ = create_subscription<graphnav_msgs::msg::NavigationGraph>(
      "~/nav_graph", 10,
      [this](const graphnav_msgs::msg::NavigationGraph::ConstSharedPtr msg) {
        planner_.update_graph(msg);
        latest_graph_header_ = msg->header;
        plan_to_goal();
      });

    goal_sub_ = create_subscription<geometry_msgs::msg::PoseStamped>(
      "~/goal_pose", 10,
      [this](const geometry_msgs::msg::PoseStamped::ConstSharedPtr msg) {
        goal_pose_ = msg;
        plan_to_goal();
      });

    odom_sub_ = create_subscription<nav_msgs::msg::Odometry>(
      "~/odom", 10, [this](const nav_msgs::msg::Odometry::ConstSharedPtr msg) { odom_ = msg; });

    path_pub_ = create_publisher<nav_msgs::msg::Path>("~/path", 10);
  }

private:
  void plan_to_goal()
  {
    if (!goal_pose_ || !latest_graph_header_) {
      return;
    }

    geometry_msgs::msg::PoseStamped goal = *goal_pose_;
    goal.header.stamp = latest_graph_header_->stamp;

    geometry_msgs::msg::PoseStamped goal_in_graph_frame;
    try {
      goal_in_graph_frame = tf_buffer_.transform(goal, latest_graph_header_->frame_id, tf2::durationFromSec(0.1));
    } catch (const tf2::TransformException& ex) {
      RCLCPP_WARN(get_logger(), "Could not transform goal pose to graph frame: %s", ex.what());
      return;
    }

    Eigen::Vector3d goal_vec(goal_in_graph_frame.pose.position.x, goal_in_graph_frame.pose.position.y, goal_in_graph_frame.pose.position.z);
    auto path = planner_.plan_to_goal(goal_vec, goal_radius_, get_clock()->now());

    nav_msgs::msg::Path path_msg;
    path_msg.header = *latest_graph_header_;
    path_msg.poses.resize(path.size());
    for (size_t i = 0; i < path.size(); ++i) {
      path_msg.poses[i].header = path_msg.header;
      path_msg.poses[i].pose.position.x = path[i].x();
      path_msg.poses[i].pose.position.y = path[i].y();
      path_msg.poses[i].pose.position.z = path[i].z();
      if (i < path.size() - 1) {
        Eigen::Vector3d d = (path[i + 1] - path[i]).normalized();
        Eigen::Matrix3d m = Eigen::Matrix3d::Identity();
        m.col(1) = Eigen::Vector3d::UnitZ().cross(d).normalized();
        m.col(0) = m.col(1).cross(m.col(2)).normalized();
        Eigen::Quaterniond q(m);
        path_msg.poses[i].pose.orientation.x = q.x();
        path_msg.poses[i].pose.orientation.y = q.y();
        path_msg.poses[i].pose.orientation.z = q.z();
        path_msg.poses[i].pose.orientation.w = q.w();
      } else {
        path_msg.poses[i].pose.orientation = goal_in_graph_frame.pose.orientation;
      }
    }
    path_pub_->publish(path_msg);

    if (odom_) {
      try {
        geometry_msgs::msg::PoseStamped goal_in_odom_frame = tf_buffer_.transform(goal, odom_->header.frame_id, tf2::durationFromSec(0.1));
        Eigen::Vector3d goal_odom(goal_in_odom_frame.pose.position.x, goal_in_odom_frame.pose.position.y, goal_in_odom_frame.pose.position.z);
        Eigen::Vector3d odom_vec(odom_->pose.pose.position.x, odom_->pose.pose.position.y, odom_->pose.pose.position.z);
        if ((goal_odom - odom_vec).norm() < goal_radius_) {
          goal_pose_.reset();
        }
      } catch (const tf2::TransformException& ex) {
        RCLCPP_WARN(get_logger(), "Could not transform goal pose to odom frame: %s", ex.what());
      }
    }
  }

  rclcpp::Subscription<graphnav_msgs::msg::NavigationGraph>::SharedPtr graph_sub_;
  rclcpp::Publisher<nav_msgs::msg::Path>::SharedPtr path_pub_;
  rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr goal_sub_;
  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_;
  tf2_ros::Buffer tf_buffer_;
  tf2_ros::TransformListener tf_listener_;
  geometry_msgs::msg::PoseStamped::ConstSharedPtr goal_pose_;
  nav_msgs::msg::Odometry::ConstSharedPtr odom_;
  std::optional<std_msgs::msg::Header> latest_graph_header_;
  double goal_radius_;
  Planner planner_;
};

}  // namespace wildos_graphnav_planner

#include "rclcpp_components/register_node_macro.hpp"
RCLCPP_COMPONENTS_REGISTER_NODE(wildos_graphnav_planner::PlannerNode)

File: src/graphnav_mpc_planner/graphnav_mpc_planner/planner_node.py
"""
GraphNav Dijkstra planner node.

Subscribes:
  ~/nav_graph   graphnav_msgs/NavigationGraph   -- scored navigation graph
  ~/goal_pose   geometry_msgs/PoseStamped       -- navigation goal
  ~/odom        nav_msgs/Odometry               -- robot odometry (used to detect goal reach)

Publishes:
  ~/path        nav_msgs/Path                   -- planned waypoint sequence

The graph planner mirrors graphnav_planner::PlannerNode (C++) in Python,
replacing the C++ PathFollowerNode with MPC downstream.

author: Lorenzo Ortolani
"""

import math

import rclpy
from rclpy.node import Node

import tf2_ros
import tf2_geometry_msgs  # noqa: F401  (registers PoseStamped transform)

from geometry_msgs.msg import PoseStamped
from graphnav_msgs.msg import NavigationGraph
from nav_msgs.msg import Odometry, Path
from std_msgs.msg import Header
from visualization_msgs.msg import MarkerArray

from graphnav_mpc_planner.graph_planner import GraphNavPlanner


class PlannerNode(Node):

    def __init__(self):
        super().__init__('graphnav_mpc_planner_node')

        # ── Parameters ────────────────────────────────────────────────
        self.declare_parameter('frontier_dist_cost_factor', 2.0)
        self.declare_parameter('goal_dist_cost_factor',     1.0)
        self.declare_parameter('frontier_score_factor',    10.0)
        self.declare_parameter('min_local_frontier_score',  0.4)
        self.declare_parameter('local_frontier_radius',     7.0)
        self.declare_parameter('path_smoothness_period',   10.0)
        self.declare_parameter('trav_class',            'default')
        self.declare_parameter('goal_radius',               3.0)

        # ── Planner ────────────────────────────────────────────────────
        self._planner = GraphNavPlanner(logger=self.get_logger())
        self._planner.frontier_dist_cost_factor = float(
            self.get_parameter('frontier_dist_cost_factor').value)
        self._planner.goal_dist_cost_factor = float(
            self.get_parameter('goal_dist_cost_factor').value)
        self._planner.frontier_score_factor = float(
            self.get_parameter('frontier_score_factor').value)
        self._planner.min_local_frontier_score = float(
            self.get_parameter('min_local_frontier_score').value)
        self._planner.local_frontier_radius = float(
            self.get_parameter('local_frontier_radius').value)
        self._planner.path_smoothness_period = float(
            self.get_parameter('path_smoothness_period').value)
        self._planner.set_trav_class(
            self.get_parameter('trav_class').get_parameter_value().string_value)

        self._goal_radius = float(self.get_parameter('goal_radius').value)

        # ── TF ────────────────────────────────────────────────────────
        self._tf_buffer = tf2_ros.Buffer()
        self._tf_listener = tf2_ros.TransformListener(self._tf_buffer, self)

        # ── State ─────────────────────────────────────────────────────
        self._goal_pose: PoseStamped | None = None
        self._odom: Odometry | None = None
        self._latest_graph_header: Header | None = None

        # ── Subscribers ───────────────────────────────────────────────
        self.create_subscription(
            NavigationGraph, '~/nav_graph', self._graph_cb, 10)
        self.create_subscription(
            PoseStamped, '~/goal_pose', self._goal_cb, 10)
        self.create_subscription(
            Odometry, '~/odom', self._odom_cb, 10)

        # ── Publishers ────────────────────────────────────────────────
        self._path_pub = self.create_publisher(Path, '~/path', 10)

        # ── Debug state ───────────────────────────────────────────────
        self._graph_received = False
        self._goal_received = False
        self._odom_received = False

        self.get_logger().info('GraphNav MPC planner node ready')
        self.get_logger().info('Waiting for topics:')
        self.get_logger().info('  - ~/nav_graph (NavigationGraph)')
        self.get_logger().info('  - ~/goal_pose (PoseStamped)')
        self.get_logger().info('  - ~/odom (Odometry)')
        self.get_logger().info('Use `ros2 topic list` to check available topics')

    # ── Callbacks ─────────────────────────────────────────────────────

    def _graph_cb(self, msg: NavigationGraph):
        if not self._graph_received:
            self.get_logger().info(f'✓ Received navigation graph with {len(msg.nodes)} nodes')
            self._graph_received = True
        self._planner.update_graph(msg)
        self._latest_graph_header = msg.header
        self._replan()

    def _goal_cb(self, msg: PoseStamped):
        if not self._goal_received:
            self.get_logger().info(f'✓ Received goal pose in frame "{msg.header.frame_id}"')
            self._goal_received = True
        self._goal_pose = msg
        self._replan()

    def _odom_cb(self, msg: Odometry):
        if not self._odom_received:
            self.get_logger().info(f'✓ Receiving odometry from frame "{msg.header.frame_id}"')
            self._odom_received = True
        self._odom = msg
        if self._goal_pose is None:
            return
        # Clear goal once robot is within goal_radius
        try:
            odom_pose = PoseStamped()
            odom_pose.header = msg.header
            odom_pose.pose = msg.pose.pose
            goal_in_odom = self._tf_buffer.transform(
                self._goal_pose, msg.header.frame_id, timeout=rclpy.duration.Duration(seconds=0.1)
            )
            dx = goal_in_odom.pose.position.x - msg.pose.pose.position.x
            dy = goal_in_odom.pose.position.y - msg.pose.pose.position.y
            if math.hypot(dx, dy) < self._goal_radius:
                self.get_logger().info('Goal reached — clearing goal')
                self._goal_pose = None
        except Exception:
            pass

    # ── Planning ──────────────────────────────────────────────────────

    def _replan(self):
        if self._goal_pose is None or self._latest_graph_header is None:
            missing = []
            if self._goal_pose is None:
                missing.append('goal_pose')
            if self._latest_graph_header is None:
                missing.append('nav_graph')
            if missing:
                self.get_logger().debug(
                    f'Cannot plan yet — waiting for: {", ".join(missing)}', throttle_duration_sec=5.0
                )
            return

        # Transform goal into the graph coordinate frame
        try:
            goal_stamped = PoseStamped()
            goal_stamped.header.stamp = self._latest_graph_header.stamp
            goal_stamped.header.frame_id = self._goal_pose.header.frame_id
            goal_stamped.pose = self._goal_pose.pose
            goal_in_graph = self._tf_buffer.transform(
                goal_stamped,
                self._latest_graph_header.frame_id,
                timeout=rclpy.duration.Duration(seconds=0.1),
            )
        except Exception as ex:
            self.get_logger().warning(
                f'Could not transform goal to graph frame: {ex}', throttle_duration_sec=2.0
            )
            return

        goal_xyz = (
            goal_in_graph.pose.position.x,
            goal_in_graph.pose.position.y,
            goal_in_graph.pose.position.z,
        )

        now_sec = self.get_clock().now().nanoseconds * 1e-9
        path_pts = self._planner.plan_to_goal(goal_xyz, self._goal_radius, now_sec)

        if not path_pts:
            self.get_logger().warning(
                'No path found by graph planner', throttle_duration_sec=5.0
            )
            return

        # ── Build nav_msgs/Path ───────────────────────────────────────
        path_msg = Path()
        path_msg.header = self._latest_graph_header

        for i, (px, py, pz) in enumerate(path_pts):
            ps = PoseStamped()
            ps.header = self._latest_graph_header
            ps.pose.position.x = px
            ps.pose.position.y = py
            ps.pose.position.z = pz

            # Orientation: heading toward next waypoint
            if i < len(path_pts) - 1:
                dx = path_pts[i + 1][0] - px
                dy = path_pts[i + 1][1] - py
                yaw = math.atan2(dy, dx)
            else:
                yaw = _quat_to_yaw(
                    goal_in_graph.pose.orientation.x,
                    goal_in_graph.pose.orientation.y,
                    goal_in_graph.pose.orientation.z,
                    goal_in_graph.pose.orientation.w,
                )
            ps.pose.orientation = _yaw_to_quat_msg(yaw)
            path_msg.poses.append(ps)

        self._path_pub.publish(path_msg)


# ─────────────────────────────────────────────────────────────────────────────
# Utilities
# ─────────────────────────────────────────────────────────────────────────────

def _quat_to_yaw(qx: float, qy: float, qz: float, qw: float) -> float:
    siny = 2.0 * (qw * qz + qx * qy)
    cosy = 1.0 - 2.0 * (qy * qy + qz * qz)
    return math.atan2(siny, cosy)


def _yaw_to_quat_msg(yaw: float):
    from geometry_msgs.msg import Quaternion
    half = yaw / 2.0
    q = Quaternion()
    q.x = 0.0
    q.y = 0.0
    q.z = math.sin(half)
    q.w = math.cos(half)
    return q


# ─────────────────────────────────────────────────────────────────────────────

def main(args=None):
    rclpy.init(args=args)
    node = PlannerNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        try:
            rclpy.shutdown()
        except Exception:
            pass


if __name__ == '__main__':
    main()
